src/devices/Pca95x4/Pca95x4.cs

```diff
@@ -0,0 +1,199 @@
+        /// </summary>
+        /// <param name="i2cDevice">The I2C device used for communication.</param>
+        /// <param name="interrupt">The input pin number that is connected to the interrupt (INT).</param>
+        public Pca95x4(I2cDevice i2cDevice, int? interrupt = null)
+        {
+            _i2cDevice = i2cDevice;
+            _interrupt = interrupt;
+
+            InitializeMasterGpioController();
+        }
+
+        private static void ValidateBitNumber(int bitNumber)
+        {
+            if (bitNumber < 0 || bitNumber > 7)
+            {
+                throw new IndexOutOfRangeException($"Invalid bit number {bitNumber}.");
+            }
+        }
+
+        private static void ClearBit(ref byte data, int bitNumber)
+        {
+            ValidateBitNumber(bitNumber);
+            data &= (byte)~(1 << bitNumber);
+        }
+
+        private void SetBit(ref byte data, int bitNumber)
+        {
+            ValidateBitNumber(bitNumber);
+            data |= (byte)(1 << bitNumber);
+        }
+
+        private static bool GetBit(byte data, int bitNumber)
+        {
+            ValidateBitNumber(bitNumber);
+            return ((data >> bitNumber) & 1) == 1;
+        }
+
+        private void InitializeMasterGpioController()
+        {
+            // Only need master controller if there is external pin provided.
+            if (_interrupt != null)
+            {
+                _masterGpioController = new GpioController();
+                _masterGpioController.OpenPin((int)_interrupt, PinMode.Input);
+            }
+        }
+
+        /// <summary>
+        /// Reads a byte from a register.
+        /// </summary>
+        /// <param name="register">The register to read.</param>
+        /// <returns>The data read from the register.</returns>
+        public byte Read(Register register)
+        {
+            byte[] data = Read(register, 1);
+            return data[0];
+        }
+
+        /// <summary>
+        /// Reads a number of bytes from a register. 
+        /// All data returned is from the selected register as Pca95x4 does not auto-increment addressing.
+        /// </summary>
+        /// <param name="register">The register to read.</param>
+        /// <param name="byteCount"></param>
+        /// <returns>The data read from the register.</returns>
+        public byte[] Read(Register register, byte byteCount)
+        {
+            Write(register, new byte[] { }); // Set address to register first.
+
+            byte[] readBuffer = new byte[byteCount];
+            _i2cDevice.Read(readBuffer);
+            return readBuffer;
+        }
+
+        /// <summary>
+        /// Reads a bit from a register.
+        /// </summary>
+        /// <param name="register">The register to read.</param>
+        /// <param name="bitNumber">The register bit number to read.</param>
+        /// <returns>The value of the register bit read.</returns>
+        public bool ReadBit(Register register, int bitNumber)
```

Comment:
This guy is not that far off from the API that I suggested I believe. Given that the main point of this device is to provide pins, I think it would be useful to have a ReadPin (instead of ReadBit) where you read from the input register, and a WritePin where you write to either the output register or the configuration one.

Reply:
I'm hesitant to add Reading/Writing pin for I/O (except for control lines; reset, interrupts, etc.) right now.

1. While I think the Mcp23xxx will be the more popular I/O expander peeps will use, I worked on this one as an exercise to see other type of approaches/circuity other expanders have.  Pca95x4 is basic compared to the Mcp23xxx, but one reason is this binding only includes the 8-bit (1 port).  There are others in related family, but the internal circuitry is different and I didn't see a good solution like we did with Mcp23xxx.
2. Since we used the ReadBit/WriteBit with Mcp23xxx, I was trying to be a little consistent.  This isn't a hard requirement, though.
3. I can see ReadPin since there is the 1 port, but a user might want to read bits from the other 3 ports, as well.  So ReadPin wouldn't make sense as much.  Unless you meant have ReadPin in addition to ReadBit.  WritePin to the Polarity register would be a little confusing, too.
4. The main reason I still think it useful to separate pin commanding (hence part of the #125 proposal), I can see users getting configurations in an incorrect state when mixing controlling pins and updating register data.  For example, let's say I have a binding that included I/O direction state and set the odd pins to be outputs.  Then you directly write to the register that changes an I/O to an input.  You could update the state at the same time, but gets a little more complex as you have to start adding if statements throughout checking if they used specific registers.  This doesn't account for pull-up/down, drain, collector mode types either.

Sorry for the long post on this.  I just haven't figured out a good story for this yet.  I'll keep pondering.